### PR TITLE
e2e: v8.1 upgrade test for capital efficient escrows

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

closes: #5635 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: e2e/tests/upgrades/upgrade_test.go [hunk 1]
```diff
@@ -59,8 +62,12 @@ func (s *UpgradeTestSuite) UpgradeChain(ctx context.Context, chain *cosmos.Cosmo
...
 	}
 
-	upgradeProposal := upgradetypes.NewSoftwareUpgradeProposal(fmt.Sprintf("upgrade from %s to %s", currentVersion, upgradeVersion), "upgrade chain E2E test", plan)
-	s.ExecuteAndPassGovV1Beta1Proposal(ctx, chain, wallet, upgradeProposal)
+	msgSoftwareUpgrade := &upgradetypes.MsgSoftwareUpgrade{
+		Plan:      plan,
+		Authority: authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+	}
+
+	s.ExecuteAndPassGovV1Proposal(ctx, msgSoftwareUpgrade, chain, wallet)
```
Comment (damiannolan) [inline e2e/tests/upgrades/upgrade_test.go:70]: I guess we should add some compatibility support here, then should be ready for review

File: e2e/tests/upgrades/upgrade_test.go [hunk 2]
```diff
@@ -59,8 +62,12 @@ func (s *UpgradeTestSuite) UpgradeChain(ctx context.Context, chain *cosmos.Cosmo
 		Info:   fmt.Sprintf("upgrade version test from %s to %s", currentVersion, upgradeVersion),
 	}
 
-	upgradeProposal := upgradetypes.NewSoftwareUpgradeProposal(fmt.Sprintf("upgrade from %s to %s", currentVersion, upgradeVersion), "upgrade chain E2E test", plan)
-	s.ExecuteAndPassGovV1Beta1Proposal(ctx, chain, wallet, upgradeProposal)
```
Comment (damiannolan) [inline e2e/tests/upgrades/upgrade_test.go:63]: curious when it's possible to completely switch over and remove this

File: e2e/tests/upgrades/upgrade_test.go [hunk 3]
```diff
@@ -606,6 +613,136 @@ func (s *UpgradeTestSuite) TestV7ToV8ChainUpgrade() {
...
 
+func (s *UpgradeTestSuite) TestV8ToV8_1ChainUpgrade() {
+	t := s.T()
+
+	ctx := context.Background()
+	relayer, channelA := s.SetupChainsRelayerAndChannel(ctx, func(opts *ibc.CreateChannelOptions) {
+		opts.Version = "{\"fee_version\":\"ics29-1\",\"app_version\":\"ics20-1\"}"
+		opts.DestPortName = "transfer"
+		opts.SourcePortName = "transfer"
+	})
```
Comment (damiannolan) [inline e2e/tests/upgrades/upgrade_test.go:624]: could import `transfertypes` and use consts here, maybe format the version. I copied this from one of the fee tests directly

File: e2e/tests/upgrades/upgrade_test.go [hunk 4]
```diff
@@ -606,6 +613,136 @@ func (s *UpgradeTestSuite) TestV7ToV8ChainUpgrade() {
...
+
+	t.Run("29-fee migration partially refunds escrowed tokens", func(t *testing.T) {
+		actualBalance, err := s.GetChainANativeBalance(ctx, chainAWallet)
+		s.Require().NoError(err)
+
+		testFee := testvalues.DefaultFee(chainADenom)
+		legacyTotal := testFee.RecvFee.Add(testFee.AckFee...).Add(testFee.TimeoutFee...)
+		refundCoins := legacyTotal.Sub(testFee.Total()...) // Total() returns the denomwise max of (recvFee + ackFee, timeoutFee)
+
+		expected := testvalues.StartingTokenAmount - testvalues.IBCTransferAmount - legacyTotal.AmountOf(chainADenom).Int64() + refundCoins.AmountOf(chainADenom).Int64()
```
Comment (damiannolan) [inline e2e/tests/upgrades/upgrade_test.go:692]: can break this down further if it makes sense, into `prevBalance` or something

File: .github/workflows/e2e-upgrade.yaml [hunk 5]
```diff
@@ -12,20 +12,6 @@ on:
 
 
 jobs:
-  upgrade-v5-hermes:
```
Comment (damiannolan) [inline .github/workflows/e2e-upgrade.yaml:15]: EOL btw, so removed both hermes and rly tests

File: e2e/tests/upgrades/upgrade_test.go (hunk 3 above)
Comment (DimitrisJim) [reply e2e/tests/upgrades/upgrade_test.go:624]: could maybe move it into `testsuite` as is done for `TransferChannelOptions` if its used by both but fine with either, should fail if something changes.

File: e2e/testsuite/testconfig.go [hunk 6]
```diff
@@ -723,3 +731,17 @@ func modifyClientGenesisAppState(ibcAppState []byte) ([]byte, error) {
 
 	return ibcGenBz, nil
 }
+
+// modifyChannelGenesisAppState takes the existing ibc app state and marshals it to a ibc GenesisState.
```
Comment (DimitrisJim) [inline e2e/testsuite/testconfig.go:735]: want to add shenanigans you do in function to docstring? Its same as other modify one, might be nice to drop a line here

File: e2e/tests/upgrades/upgrade_test.go (hunk 2 above)
Comment (crodriguezvega) [reply e2e/tests/upgrades/upgrade_test.go:63]: That would be once v7 reaches end of life, right?

File: e2e/testsuite/testsuite.go [hunk 7]
```diff
@@ -399,6 +400,22 @@ func (*E2ETestSuite) TransferChannelOptions() func(options *ibc.CreateChannelOpt
 	}
 }
 
+// FeeMiddlewareChannelOptions configures both of the chains to have fee middleware enabled.
+func (s *E2ETestSuite) FeeMiddlewareChannelOptions() func(options *ibc.CreateChannelOptions) {
```
Comment (crodriguezvega) [inline e2e/testsuite/testsuite.go:404]: I took the liberty to move this function here.

**Review by crodriguezvega — APPROVED** [review]
Great work! I pushed some changes to address a couple of review comments, and the [tests still pass](https://github.com/cosmos/ibc-go/actions/runs/7603675687), so giving my approval now.

File: e2e/testsuite/testsuite.go (hunk 7 above)
Comment (damiannolan) [reply e2e/testsuite/testsuite.go:404]: Thank you @crodriguezvega!

File: e2e/tests/upgrades/upgrade_test.go (hunk 2 above)
Comment (damiannolan) [reply e2e/tests/upgrades/upgrade_test.go:63]: I think so, but I'm basing it just off one of the feature releases matrices in values.go

File: e2e/testsuite/testconfig.go [hunk 8]
```diff
@@ -723,3 +731,18 @@ func modifyClientGenesisAppState(ibcAppState []byte) ([]byte, error) {
...
+
+// modifyChannelGenesisAppState takes the existing ibc app state, unmarshals it to a map and removes the `params` entry from ibc channel genesis.
+// It marshals and returns the ibc GenesisState JSON map as bytes.
+func modifyChannelGenesisAppState(ibcAppState []byte) ([]byte, error) {
+	var ibcGenesisMap map[string]interface{}
+	if err := json.Unmarshal(ibcAppState, &ibcGenesisMap); err != nil {
+		return nil, err
+	}
+
+	// be ashamed, be very ashamed
```
Comment (chatton) [inline e2e/testsuite/testconfig.go:743]: :heart: 

This is not actually the worst thing ATM, I think we can improve this, but this ended up being very readable and concise compared to the other options we were playing with.

**Review by chatton — APPROVED** [review]
LGTM! Thank you for some of the workflow cleanup and sorting out the channel params genesis issue :heart:

File: e2e/testsuite/testconfig.go (hunk 8 above)
Comment (damiannolan) [reply e2e/testsuite/testconfig.go:743]: yea, agree! let's leave in the comment for comical effect I guess, would be nice to not have to add a new func for every field we need to sanitise, but we can think about that later - string splitting on a `jsonPath` with "." is a bit gross but I'm not sure I have any better ideas atm!

**Pull request merged**